### PR TITLE
fix issue for aws iam profile [bsc#1169506] (1517)

### PR DESCRIPTION
**Pull request by cclhsu**
Fix issue items for aws iam profile [bsc#1169506]

Signed-off-by: cclhsu <clark.hsu@suse.com>

## Why is this PR needed?

Fix issue items for aws iam profile [bsc#1169506]

Fixes #1517

## What does this PR do?

Create aws_iam_role and aws_iam_policy

## Anything else a reviewer needs to know?

Special test cases, manual steps, links to resources or anything else that could be helpful to the reviewer.

## Info for QA

This is info for QA so that they can validate this. This is **mandatory** if this PR fixes a bug.
If this is a new feature, a good description in "What does this PR do" may be enough.

### Related info

Info that can be relevant for QA:
* link to other PRs that should be merged together
* link to packages that should be released together
* upstream issues

### Status **BEFORE** applying the patch

How can we reproduce the issue? How can we see this issue? Please provide the steps and the prove
this issue is not fixed.

In terraform.tfvar, comment out following:

```
# iam_profile_master = "caasp-k8s-master-vm-profile"
# iam_profile_worker = "caasp-k8s-worker-vm-profile"
```

Deploy AWS in non-openbare account will not success.

### Status **AFTER** applying the patch

How can we validate this issue is fixed? Please provide the steps and the prove this issue is fixed.

Deploy AWS in non-openbare account will be successful.

## Docs

If docs need to be updated, please add a link to a PR to https://github.com/SUSE/doc-caasp.
At the time of creating the issue, this PR can be work in progress (set its title to [WIP]),
but the documentation needs to be finalized before the PR can be merged. 

# Merge restrictions

(Please do not edit this)

We are in *v4-maintenance phase*, so we will restrict what can be merged to prevent unexpected surprises:

    What can be merged (merge criteria):
        2 approvals:
            1 developer: code is fine
            1 QA: QA is fine
        there is a PR for updating documentation (or a statement that this is not needed)

### CODE REVIEW

File: ci/infra/aws/master-instance.tf [hunk 1]
```diff
@@ -7,11 +7,11 @@ resource "aws_instance" "control_plane" {
   source_dest_check           = false
   subnet_id                   = aws_subnet.public.id
   user_data                   = data.template_cloudinit_config.cfg.rendered
-  iam_instance_profile        = length(var.iam_profile_master) == 0 ? local.aws_iam_policy_master_terraform : var.iam_profile_master
+  iam_instance_profile        = length(var.iam_profile_master) == 0 ? local.aws_iam_role_policy_master_terraform : var.iam_profile_master
 
   depends_on = [
     aws_internet_gateway.platform,
-    aws_iam_policy.master,
+    aws_iam_role_policy.master,
```
Comment (innobead) [inline ci/infra/aws/master-instance.tf:14]: aws_iam_instance_profile.master instead?

File: ci/infra/aws/worker-instance.tf [hunk 2]
```diff
@@ -6,12 +6,12 @@ resource "aws_instance" "nodes" {
...
   source_dest_check           = false
   user_data                   = data.template_cloudinit_config.cfg.rendered
-  iam_instance_profile        = length(var.iam_profile_worker) == 0 ? local.aws_iam_policy_worker_terraform : var.iam_profile_worker
+  iam_instance_profile        = length(var.iam_profile_worker) == 0 ? local.aws_iam_role_policy_worker_terraform : var.iam_profile_worker
   subnet_id                   = aws_subnet.private.id
 
   depends_on = [
     aws_route.private_nat_gateway,
-    aws_iam_policy.worker,
+    aws_iam_role_policy.worker,
```
Comment (innobead) [inline ci/infra/aws/worker-instance.tf:14]: aws_iam_instance_profile.worker instead?

File: ci/infra/aws/iam_policies.tf [hunk 3]
```diff
@@ -2,15 +2,43 @@
 # The expected policies are defined here: https://github.com/kubernetes/cloud-provider-aws#readme
 
 locals {
-  aws_iam_policy_master_terraform = "${var.stack_name}_cpi_master_policy"
-  aws_iam_policy_worker_terraform = "${var.stack_name}_cpi_worker_policy"
+  aws_iam_role_policy_master_terraform = "${var.stack_name}_cpi_master_policy"
```
Comment (innobead) [inline ci/infra/aws/iam_policies.tf:5]: aws_iam_instance_profile_master_terraform = "${var.stack_name}_cpi_master"

File: ci/infra/aws/iam_policies.tf [hunk 4]
```diff
@@ -2,15 +2,43 @@
...
+  aws_iam_role_policy_worker_terraform = "${var.stack_name}_cpi_worker_policy"
```
Comment (innobead) [inline ci/infra/aws/iam_policies.tf:6]: aws_iam_instance_profile_worker_terraform = "${var.stack_name}_cpi_worker"

**Comment by c3y1huang** [comment]
Is there a test case for this fix?

**Comment by innobead** [comment]
> Is there a test case for this fix?

Good ask ;) @cclhsu please add a test case in `caasp-test-cases`. cc @maximenoel8

**Comment by jenting** [comment]
I don't have an account with permission to create an IAM role to do the test, so personally, I'd prefer let QA testing this PR and approve it.

File: ci/infra/aws/master-instance.tf (hunk 1 above)
Comment (cclhsu) [reply ci/infra/aws/master-instance.tf]: changed and tested.

File: ci/infra/aws/worker-instance.tf (hunk 2 above)
Comment (cclhsu) [reply ci/infra/aws/worker-instance.tf]: changed and tested.

File: ci/infra/aws/iam_policies.tf (hunk 3 above)
Comment (cclhsu) [reply ci/infra/aws/iam_policies.tf]: changed and tested.

File: ci/infra/aws/iam_policies.tf (hunk 4 above)
Comment (cclhsu) [reply ci/infra/aws/iam_policies.tf]: changed and tested.

**Comment by zoopster** [comment]
These changes worked for me as well.

**Comment by cclhsu** [comment]
Have tested after merged. Deployed successfully

**Comment by maximenoel8** [comment]
Have tested the deployment with and without iam_profile. It's working fine for me

**Pull request merged**